shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,317 @@
+            LEAVING,
+        };
+
+        /**
+         * Message used to tell the Director about a new Provides reaction
+         */
+        struct ProvidesReaction {
+            ProvidesReaction(const std::shared_ptr<NUClear::threading::Reaction>& reaction,
+                             const std::type_index& type,
+                             const ProviderAction& action)
+                : reaction(reaction), type(type), action(action) {}
+
+            /// The reaction for this provider
+            std::shared_ptr<NUClear::threading::Reaction> reaction;
+            /// The data type that this provider services
+            std::type_index type;
+            /// The action type that this provider is using (normal, entering, leaving)
+            ProviderAction action;
+        };
+
+        /**
+         * Message used to tell the Director about a new When condition on a provider
+         */
+        struct WhenExpression {
+            WhenExpression(const std::shared_ptr<NUClear::threading::Reaction>& reaction,
+                           const std::type_index& type,
+                           bool (*validator)(const int&),
+                           int (*current)())
+                : reaction(reaction), type(type), validator(validator), current(current) {}
+
+            /// The provider reaction that is contingent on this when condition
+            std::shared_ptr<NUClear::threading::Reaction> reaction;
+            /// The enum type that this when expression is looking at
+            std::type_index type;
+            /// Function to determine if the passed state is valid
+            bool (*validator)(const int&);
+            /// Function to get the current state from the global cache
+            int (*current)();
+            /// Function to bind a reaction to monitor when this state changes
+            NUClear::threading::ReactionHandle (*binder)(NUClear::Reactor&, std::function<void(const int&)>);
+        };
+
+        /**
+         * Message used to tell the director about a new Causing relationship for this provider
+         */
+        struct CausingExpression {
+            CausingExpression(const std::shared_ptr<NUClear::threading::Reaction>& reaction,
+                              const std::type_index& type,
+                              const int& resulting_state)
+                : reaction(reaction), type(type), resulting_state(resulting_state) {}
+            /// The provider reaction that will cause this state
+            std::shared_ptr<NUClear::threading::Reaction> reaction;
+            /// The enum type that this when expression claiming to manipulate
+            std::type_index type;
+            /// The resulting state this provider is claiming to provide if it is executed
+            int resulting_state;
+        };
+
+        /**
+         * Message to tell the Director that a provider task has finished executing
+         */
```

Comment:
Have you considered making this a `proto` message? It seems like there's nothing too fancy about it, and it would fit that mould

Reply:
When writing extensions I keep the messages that will go between the extension and the module that is processing for the extension together.
Nobody should ever be emitting these particular messages outside of this exact file, to do so would be an error.
Therefore I don't want to make it a public message. If I could I would hide them even more than I have by putting them in a subnamespace here.

Reply:
So the decision against proto is purely for encapsulation?

Reply:
And because we don't want other modules actually thinking these are things that should be emitted.